game/src/Player/PassThrough.gd

```diff
@@ -0,0 +1,19 @@
+extends Area2D
+
+const PASS_THROUGH_LAYER = 3
+
+func _ready() -> void:
+	connect("body_exited", self, "_on_body_exited")
+
+
+func _on_body_exited(body: CollisionObject2D) -> void:
```

Comment:
Do you still need an area? As you're toggling the pass through collision bit on/off upon pressing/releasing the down key?

Reply:
The `_body_entered` callback is still necessary to automatically detect when the player passed through a platform, and toggle the collision bit again, preventing the player from passing through multiple platforms. If the later is desirable, then this can be a simple bit of code in the `Move` state.

Reply:
Makes sense. So the node should stay. You should only change the code so this doesn't mess with the state machine directly. A state could use this node instead through the owner, and this node have some public methods instead of working the way it does now.

Reply:
How about the signal approach? It has a signal that tells when the pass-through layers was disabled, the `Move` receives this signal and changes the state to Move/Air.

Reply:
If you're using a state machine, only the states should mess with the host node. You shouldn't have other classes changing the Player's properties. Using an FSM, you know that only the active state can manipulate the host node (or the state and its parents in the case of an HFSM, but it's explicit from the state, that calls `parent.callback` directly)

So to me, the code that changes the collision layer should be in a state. It's fine if all the PassThroughDetector does is detecting when it went out of a passthrough object. This way it would also not be tied to any other node, like LedgeDetector, WallDetector...

Reply:
Understood, I will come back with a new implementation soon.

Reply:
Thanks!

Reply:
Alright, I moved the pass-through logic to the Move state instead, connecting the `body_exited` to it.